### ISSUE TITLE
Could not find declaration

### ISSUE THREAD

**Issue by jnfoster** [p4lang/p4c#934]
Compiling this program
```p4
/* -*- P4_16 -*- */
#include <core.p4>
#include <v1model.p4>

/*************************************************************************
******************************* B A S I C ********************************
*************************************************************************/

/* Headers */
typedef bit<9>  egressSpec_t;
typedef bit<48> macAddr_t;
typedef bit<32> ip4Addr_t;

header ethernet_t {
    macAddr_t dstAddr;
    macAddr_t srcAddr;
    bit<16>   etherType;
}

header ipv4_t {
    bit<4>    version;
    bit<4>    ihl;
    bit<8>    diffserv;
    bit<16>   totalLen;
    bit<16>   identification;
    bit<3>    flags;
    bit<13>   fragOffset;
    bit<8>    ttl;
    bit<8>    protocol;
    bit<16>   hdrChecksum;
    ip4Addr_t srcAddr;
    ip4Addr_t dstAddr;
}

struct basic_metadata {
    /* empty */
}

struct basic_headers {
    ethernet_t   ethernet;
    ipv4_t       ipv4;
}

/* Parser */
parser BasicParser(packet_in packet,
                out basic_headers hdr,
                inout basic_metadata meta,
                inout standard_metadata_t standard_metadata) {

    state start {
        transition parse_ethernet;
    }

    state parse_ethernet {
        packet.extract(hdr.ethernet);
        transition select(hdr.ethernet.etherType) {
            TYPE_IPV4: parse_ipv4;
            default: accept;
        }
    }

    state parse_ipv4 {
        packet.extract(hdr.ipv4);
        transition accept;
    }

}

/* VerifyChecksum */
control BasicVerifyChecksum(in basic_headers hdr, inout basic_metadata meta) {   
    apply {  }
}

/* Ingress */
control BasicIngress(inout basic_headers hdr,
                  inout basic_metadata meta,
                  inout standard_metadata_t standard_metadata) {
    action drop() {
        mark_to_drop();
    }
    
    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
        standard_metadata.egress_spec = port;
        hdr.ethernet.srcAddr = hdr.ethernet.dstAddr;
        hdr.ethernet.dstAddr = dstAddr;
        hdr.ipv4.ttl = hdr.ipv4.ttl - 1;
    }
    
    table ipv4_lpm {
        key = {
            hdr.ipv4.dstAddr: lpm;
        }
        actions = {
            ipv4_forward;
            drop;
            NoAction;
        }
        size = 1024;
        default_action = NoAction();
    }
    
    apply {
        if (hdr.ipv4.isValid()) {
            ipv4_lpm.apply();
        }
    }
}

/* Egress */
control BasicEgress(inout basic_headers hdr,
                    inout basic_metadata meta,
                    inout standard_metadata_t standard_metadata) {
    apply {  }
}

/* ComputeChecksum */
control BasicComputeChecksum(inout basic_headers  hdr, inout basic_metadata meta) {
    apply { } 
}

/* Deparser */
control BasicDeparser(packet_out packet, in basic_headers hdr) {
    apply {
        packet.emit(hdr.ethernet);
        packet.emit(hdr.ipv4);
    }
}

/*************************************************************************
****************************** M R I *************************************
*************************************************************************/

const bit<8>  UDP_PROTOCOL = 0x11;
const bit<16> TYPE_IPV4 = 0x800;
const bit<5>  IPV4_OPTION_MRI = 31;

#define MAX_HOPS 9

typedef bit<32> switchID_t;
typedef bit<32> qdepth_t;

/* Headers */
header ipv4_option_t {
    bit<1> copyFlag;
    bit<2> optClass;
    bit<5> option;
    bit<8> optionLength;
}

header mri_t {
    bit<16>  count;
}

header switch_t {
    switchID_t  swid;
    qdepth_t    qdepth;
}

struct ingress_metadata_t {
    bit<16>  count;
}

struct parser_metadata_t {
    bit<16>  remaining;
}

struct mri_metadata {
    ingress_metadata_t   ingress_metadata;
    parser_metadata_t   parser_metadata;
}

struct mri_headers {
    ipv4_option_t      ipv4_option;
    mri_t              mri;
    switch_t[MAX_HOPS] swtraces;
}

error { IPHeaderTooShort }

/* Parser */
parser MRIParser(packet_in packet,
                 in basic_headers basic_hdr,
                 out mri_headers hdr,
                 inout mri_metadata meta,
                 inout standard_metadata_t standard_metadata) {

    state start {
        verify(basic_hdr.ipv4.ihl >= 5, error.IPHeaderTooShort);
        transition select(basic_hdr.ipv4.ihl) {
            5             : accept;
            default       : parse_ipv4_option;
        }
    }

    state parse_ipv4_option {
        packet.extract(hdr.ipv4_option);
        transition select(hdr.ipv4_option.option) {
            IPV4_OPTION_MRI: parse_mri;
            default: accept;
        }
    }

    state parse_mri {
        packet.extract(hdr.mri);
        meta.parser_metadata.remaining = hdr.mri.count;
        transition select(meta.parser_metadata.remaining) {
            0 : accept;
            default: parse_swtrace;
        }
    }

    state parse_swtrace {
        packet.extract(hdr.swtraces.next);
        meta.parser_metadata.remaining = meta.parser_metadata.remaining  - 1;
        transition select(meta.parser_metadata.remaining) {
            0 : accept;
            default: parse_swtrace;
        }
    }    
}

/* Verify Checksum */
control MRIVerifyChecksum(in mri_headers hdr, inout mri_metadata meta) {   
    apply { }
}

/* Ingress */
control MRIIngress(inout mri_headers hdr,
                  inout mri_metadata meta,
                  inout standard_metadata_t standard_metadata) {
    apply { }
}

/* Egress */
control MRIEgress(inout basic_headers basic_hdr,
                  inout mri_headers hdr,
                  inout mri_metadata meta,
                  inout standard_metadata_t standard_metadata) {
    action add_swtrace(switchID_t swid) { 
        hdr.mri.count = hdr.mri.count + 1;
        hdr.swtraces.push_front(1);
        hdr.swtraces[0].swid = swid;
        hdr.swtraces[0].qdepth = (qdepth_t)standard_metadata.deq_qdepth;

        basic_hdr.ipv4.ihl = basic_hdr.ipv4.ihl + 2;
        hdr.ipv4_option.optionLength = hdr.ipv4_option.optionLength + 8; 
    }

    table swtrace {
        actions = { 
	    add_swtrace; 
	    NoAction; 
        }
        default_action = NoAction();      
    }
    
    apply {
        if (hdr.mri.isValid()) {
            swtrace.apply();
        }
    }
}

/* ComputeChecksum */
control MRIComputeChecksum(inout mri_headers hdr, inout mri_metadata meta) {
    apply { }
}
/* Deparser */
control MRIDeparser(packet_out packet, in mri_headers hdr) {
    apply {
        packet.emit(hdr.ipv4_option);
        packet.emit(hdr.mri);
        packet.emit(hdr.swtraces);                 
    }
}

/*************************************************************************
*************************** C O M B I N E D ******************************
*************************************************************************/

/* Headers */
struct metadata {
  basic_metadata basic;  
  mri_metadata mri;
}

struct headers {
  basic_headers basic;  
  mri_headers mri;
}

/* Parser */
parser MyParser(packet_in packet,
                out headers hdr,
                inout metadata meta,
                inout standard_metadata_t standard_metadata) {
    BasicParser() basic;
    MRIParser() mri;

    state start {
        basic.apply(packet, hdr.basic, meta.basic, standard_metadata);
        mri.apply(packet, hdr.basic, hdr.mri, meta.mri, standard_metadata);
        transition accept;
    }
}

/* Ingress */
control MyIngress(inout headers hdr,
                  inout metadata meta,
                  inout standard_metadata_t standard_metadata) {
    BasicIngress() basic;
    MRIIngress() mri;
    apply {
        basic.apply(hdr.basic, meta.basic, standard_metadata);
        mri.apply(hdr.mri, meta.mri, standard_metadata);
    }
}  

/* VerifyChecksum */
control MyVerifyChecksum(in headers hdr, inout metadata meta) {
    BasicVerifyChecksum() basic;
    MRIVerifyChecksum() mri;    
    apply {
        basic.apply(hdr.basic, meta.basic);
        mri.apply(hdr.mri, meta.mri);
    }
}

/* Egress */
control MyEgress(inout headers hdr,
                 inout metadata meta,
                 inout standard_metadata_t standard_metadata) {
    BasicEgress() basic;
    MRIEgress() mri;
    apply {
        basic.apply(hdr.basic, meta.basic, standard_metadata);
        mri.apply(hdr.basic, hdr.mri, meta.mri, standard_metadata);
    }
}

/* ComputeChecksum */
control MyComputeChecksum(inout headers hdr, inout metadata meta) {
    BasicComputeChecksum() basic;
    MRIComputeChecksum() mri;    
    apply {
        basic.apply(hdr.basic, meta.basic);
        mri.apply(hdr.mri, meta.mri);
    }
}

/* Deparser */
control MyDeparser(packet_out packet, in headers hdr) {
    BasicDeparser() basic;
    MRIDeparser() mri;    
    apply {
        basic.apply(packet, hdr.basic);
        mri.apply(packet, hdr.mri);
    }
}


V1Switch(
MyParser(),
MyVerifyChecksum(),
MyIngress(),
MyEgress(),
MyComputeChecksum(),
MyDeparser()
) main;
```
produces an error:
```
compose.p4(124): error: Could not find declaration for packet
        packet.emit(hdr.ethernet);
        ^^^^^^
compose.p4(125): error: Could not find declaration for packet
        packet.emit(hdr.ipv4);
        ^^^^^^
compose.p4(271): error: Could not find declaration for packet
        packet.emit(hdr.ipv4_option);
        ^^^^^^
compose.p4(272): error: Could not find declaration for packet
        packet.emit(hdr.mri);
        ^^^^^^
compose.p4(273): error: Could not find declaration for packet
        packet.emit(hdr.swtraces);
        ^^^^^^
```
This seems wrong. In all cases, `packet` is declared as a parameter to the surrounding control.

Indeed, if I change the final switch declaration to 
```p4
V1Switch(
BasicParser(),
BasicVerifyChecksum(),
BasicIngress(),
BasicEgress(),
BasicComputeChecksum(),
BasicDeparser()
) main;
```
then it works just fine.

**Comment by mbudiu-vmw** [comment]
This is a duplicate of #794

**Comment by mbudiu-vmw** [comment]
We have closed #794, so this one will be closed too.
